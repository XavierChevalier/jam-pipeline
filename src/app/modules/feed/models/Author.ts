import { InferType, object, string } from 'yup'

export const Author = object({
  name: string().required(),
  avatar: string().required().url(),
})

export type Author = InferType<typeof Author>
