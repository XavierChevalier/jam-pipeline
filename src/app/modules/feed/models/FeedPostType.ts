export enum AvailableFeedPostType {
  userShareSong = 'userShareSong',
  userShareAlbum = 'userShareAlbum',
  userSharePlaylist = 'userSharePlaylist',
  artistReleaseSong = 'artistReleaseSong',
  artistReleaseAlbum = 'artistReleaseAlbum',
}
