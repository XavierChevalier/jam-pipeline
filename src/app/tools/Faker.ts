export const generateImageUrl = (size = 50) => `https://picsum.photos/${size}`
