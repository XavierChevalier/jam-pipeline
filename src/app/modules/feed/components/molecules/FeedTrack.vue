<script setup lang="ts">
  import { computed, PropType } from 'vue'
  import IconLike from '@/app/components/molecules/AppIconLike.vue'
  import { formatDuration } from '@/app/tools/date/FormatDuration'
  import { Track } from '@/app/modules/track/models/Track'
  import { isPropertyValid } from '@/app/tools/component-properties/PropertyValidator'

  const props = defineProps({
    track: {
      type: Object as PropType<Track>,
      required: true,
      validator: isPropertyValid(Track),
    },
  })

  const trackDuration = computed(() => formatDuration(props.track.durationMs))
</script>

<template>
  <div class="bg-white-dark rounded p-2 flex">
    <img class="rounded w-16 h-16" :src="track.cover" alt="The track cover" />

    <div class="flex flex-col ml-2 justify-center">
      <span class="font-bold">{{ track.title }}</span>
      <span v-for="band in track.bands" :key="band">{{ band }}</span>
    </div>

    <div class="flex self-center ml-auto">
      <IconLike class="text-black-light" />
      <span class="ml-3">{{ trackDuration }}</span>
    </div>
  </div>
</template>
