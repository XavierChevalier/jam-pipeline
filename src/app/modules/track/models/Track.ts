import { TrackCover } from '@/app/modules/track/models/TrackCover'
import { array, InferType, number, object, string } from 'yup'

export const TrackTitle = string().required()
export const TrackBands = array(string().required()).required().min(1)

export const Track = object({
  cover: TrackCover,
  title: TrackTitle,
  bands: TrackBands,
  durationMs: number().positive().required(),
})
export type Track = InferType<typeof Track>
