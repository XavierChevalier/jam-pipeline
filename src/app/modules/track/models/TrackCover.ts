import { TrackBands, TrackTitle } from '@/app/modules/track/models/Track'
import { InferType, object, string } from 'yup'

export const TrackCover = string().url().required()
export type TrackCover = InferType<typeof TrackCover>

export const TrackOfTrackCover = object({
  cover: TrackCover,
  title: TrackTitle,
  bands: TrackBands,
})
export type TrackOfTrackCover = InferType<typeof TrackOfTrackCover>
