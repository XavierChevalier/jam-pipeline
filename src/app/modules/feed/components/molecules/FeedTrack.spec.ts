import * as Stories from '@/app/modules/feed/components/molecules/FeedTrack.stories'
import { Track } from '@/app/modules/track/models/Track'
import { omit } from 'lodash-es'
import { StorybookTestsGenerator } from '@/tests/tests-generators/StorybookTestsGenerator'
import { generateImageUrl } from '@/app/tools/Faker'

describe('FeedTrack', () => {
  StorybookTestsGenerator.fromStoriesExports(Stories)
    .snapshotEachStories()
    .property<Track>('track', (property) => {
      const track = {
        cover: generateImageUrl(),
        title: 'G.O.A.T.',
        bands: ['Polyphia'],
        durationMs: 125_000,
      }
      property
        .itShouldBeDefined()
        .itShouldHaveAValidatorFunction()
        .itShouldSuccessWith(track)
        .itShouldFailWith('missing cover', omit(track, ['cover']))
        .itShouldFailWith('missing title', omit(track, ['title']))
        .itShouldFailWith('missing bands', omit(track, ['bands']))
        .itShouldFailWith('missing durationMs', omit(track, ['durationMs']))
    })
})
