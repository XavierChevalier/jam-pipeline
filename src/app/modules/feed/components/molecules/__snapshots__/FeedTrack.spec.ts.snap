// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedTrack stories snapshots [snapshot] should render the "OneBand" component story correctly 1`] = `
<div
  class="bg-white-dark rounded p-2 flex"
>
  <img
    alt="The track cover"
    class="rounded w-16 h-16"
    src="https://picsum.photos/50"
  />
  <div
    class="flex flex-col ml-2 justify-center"
  >
    <span
      class="font-bold"
    >
      G.O.A.T.
    </span>
    
    <span>
      Polyphia
    </span>
    
  </div>
  <div
    class="flex self-center ml-auto"
  >
    <anonymous-stub
      class="text-black-light"
      counter="0"
      liked="false"
      showcounter="false"
    />
    <span
      class="ml-3"
    >
      2:05
    </span>
  </div>
</div>
`;
